### ISSUE TITLE
--quiet does not silence 'x warnings generated'

### ISSUE THREAD

**Issue by jyn514** [llvm/llvm-project#47042]
|  |  |
| --- | --- |
| Bugzilla Link | [47698](https://llvm.org/bz47698) |
| Version | unspecified |
| OS | Linux |

## Extended Description 
`clang-tidy --quiet` still prints the number of warnings generated. This is annoying and fills up build logs, especially if there are many hundreds of files. I can't silence the warnings with 2>/dev/null because then I miss the actual errors, like `compile-commands.json` being missing.

Could `--quiet` also silence these notices?

```
$ cat tmp.c 
int f() {
}
$ clang-tidy --quiet tmp.c >/dev/null
2 warnings generated.
```

**Comment by AaronBallman** [comment]
I remember looking into this at one point in time and finding out that the output isn't coming from clang-tidy, it's coming from Clang itself: https://github.com/llvm/llvm-project/blob/main/clang/lib/Frontend/CompilerInstance.cpp#L1051

Weirdly (IMO), you can pass `-fno-caret-diagnostics` to Clang (not to clang-tidy) to suppress printing that information: https://godbolt.org/z/aTWeeo8rd

I think clang-tidy should probably silence this output from Clang when passed `--quiet`. WDYT @njames93?

**Comment by njames93** [comment]
Can you not pass that flag in the extra arguments after the `--` on the command line?

**Comment by carlosgalvezp** [comment]
@njames93 It can be done, but using that flag removes useful information for when there are warnings. Example:

Without flag:
```cpp
<source>:12:12: warning: comparison of integers of different signs: 'unsigned int' and 'int' [-Wsign-compare]
  return j < i;
         ~ ^ ~
1 warning generated.
Compiler returned: 0
```
With the flag:
```
<source>:12:12: warning: comparison of integers of different signs: 'unsigned int' and 'int' [-Wsign-compare]
Compiler returned: 0
```

Will this also affect clang-tidy's ability to display the extra info?

**Comment by carlosgalvezp** [comment]
> Will this also affect clang-tidy's ability to display the extra info?
It seems it doesn't, works for me! Still awkward in case someone wants to fix it.

**Comment by hsdk123** [comment]
+1

**Comment by zufuliu** [comment]
https://github.com/llvm/llvm-project/blob/c8c583071591dbc6927955e608dcd44c0bac8b9c/clang-tools-extra/clang-tidy/tool/run-clang-tidy.py#L137-L138

https://github.com/llvm/llvm-project/blob/c8c583071591dbc6927955e608dcd44c0bac8b9c/clang-tools-extra/clang-tidy/tool/run-clang-tidy.py#L360-L362

**Comment by Poldraunic** [comment]
Today I have encountered this as well. I've never used `clang-tidy` from CLI before (always relied on IDE integration), but today I was tinkering abot with integration to CI.

This message have also stumped me and I spent couple of hours trying to figure out what's going on and how to disable it. Depending on checks enabled, this will produce a huge amount of bogus warnings that user unable to act upon most of the time. If there are identifier naming checks in place, this counter most likely won't do any good due to clash C/C++ STD naming.

I confirm what Aaron said about `--quiet` option. In my mind, it should produce no output whatsoever and user should rely on return code.

I am using `run-clang-tidy.py` and out of the box it seems there is no way to pass additional arguments to `clang-tidy`. And so `-fno-caret-diagnostics` option is out. Having code snippets is nice, but not mandatory. Given iterative approach to programming, the amount of diagnostics will be low and fixing them won't be a problem that is unless a new check is introduced.

For now I will be applying patches to Python script so that it doesn't spam useless statements into my CI log.

**Comment by lursyy** [comment]
> I think Clang should have *some* way [...] even if we [...] don't expose a new f flag for this control

@AaronBallman Would you (or anyone else) mind detailing how one could allow this control without exposing a new flag? (or perhaps a new flag is justified after all)?

**Comment by AaronBallman** [comment]
clang-tidy uses the clang tooling library, so I was thinking that the `ToolInvocation` object could perhaps take an option that says "be quiet, seriously!" which gets passed along to the `CompilerInvocation` via one of the options objects. So it doesn't have to be settable from the command line, but an internal option. (Note, this is off-the-cuff, there may be details I'm missing.)

**Comment by sambishop** [comment]
A note to anyone looking for a workaround: the `warnings generated` messages appear to be the only thing written to stderr, so they can be suppressed by redirecting stderr to `/dev/null`. (Ex: `run-clang-tidy ... 2>/dev/null`)

The warning messages themselves ("`warning: comparison of integers ...`") are written to stdout, so they'll still be seen.
